### ISSUE TITLE
Add EasyAntiCheat Runtime and BattlEye Runtime

### ISSUE THREAD

**Issue by Sultech** [Heroic-Games-Launcher/HeroicGamesLauncher#1513]
### Problem description

When trying to play games like Fall guys, the game will not work because it does not have the EasyAntiCheat Runtime.

### Feature description

Add a toggle in the settings to enable EasyAntiCheat Runtime and BattlEye Runtime. This is helpful when trying to play games like Fall guys and other EAC and Battleye games that have enabled support but you need the runtimes to make them actually work.

**Comment by Nocccer** [comment]
That would be cool, but i don't know where you can get this runtimes via an api ? Does Lutris have this feature?

**Comment by Nocccer** [comment]
If Lutris has it we can probably add it.

**Comment by Etaash-mathamsetty** [comment]
> If Lutris has it we can probably add it.

for EAC, add this env variable:
PROTON_EAC_RUNTIME=PATH_TO_EAC_RUNTIME
and for battleye:
PROTON_BATTLEYE_RUNTIME=PATH_TO_BE_RUNTIME

https://github.com/lutris/lutris/blob/c4f10b24944a907e5ac0c7a89682f6ec9ace3222/lutris/runners/wine.py
obviously need a version of proton!

**Comment by GitThisBugOff** [comment]
For what it's worth, I still can't get Fall Guys to run after adding the
`PROTON_EAC_RUNTIME` variable.

![Screenshot_20220621_225347](https://user-images.githubusercontent.com/39744079/174933395-1684ea37-d5b3-467f-bbe8-5cf2b0d79575.png)

**Comment by imLinguin** [comment]
We've been looking into that on Discord channel, Fall Guys is refusing to work even with this env var.

**Comment by emmandyar** [comment]
Fall Guys works if you specify the proton EAC runtime, as described by @Etaash-mathamsetty  and symlink/copy the correct eac library in the game's install directory:

> Make a symlink or copy of FallGuys-install-dir/EasyAntiCheat/easyanticheat_x64.so in folder FallGuys-install-dir/FallGuys_client_game_Data/Plugins/x86_64/

Replace Fallguys-install-dir with the actual path (by default, /home/user/Games/Heroic/FallGuys)

**Comment by faedy2** [comment]
> Fall Guys works if you specify the proton EAC runtime, as described by @Etaash-mathamsetty and symlink/copy the correct eac library in the game's install directory:
> 
> > Make a symlink or copy of FallGuys-install-dir/EasyAntiCheat/easyanticheat_x64.so in folder FallGuys-install-dir/FallGuys_client_game_Data/Plugins/x86_64/
> 
> Replace Fallguys-install-dir with the actual path (by default, /home/user/Games/Heroic/FallGuys)
> 
> EDIT: You also need to set a custom executable:
> 
> > In Heroic's settings for the game, go to Other and designate the alternative exe as FallGuys-install-dir/FallGuys_client_game.exe. Alternatively, edit FallGuys-install-dir/FallGuys_client.ini and change the first line to be TargetApplicationPath=FallGuys_client_game.exe.

I got the game to launch!!!

**Comment by faedy2** [comment]
Welp, spoke too soon
When I try to play, the game tells me that the anti-cheat is not initialized

**Comment by sheabert33** [comment]
I too am able to get to the main menu but am unable to start the game.  I followed the workaround mentioned on the steam page but it doesn't seem to do the trick for Heroic Games.

**Comment by GitThisBugOff** [comment]
@Etaash-mathamsetty Still can't get it to detect EAC. I copied the EAC folder and placed it inside the `/Games/Heroic` folder, renamed it to something easier and the game still complains not finding it.

![image](https://user-images.githubusercontent.com/39744079/175188296-3b9d61f7-1dd9-4d86-8558-7acc068e1f7b.png)


I'm using Proton-Experimental.

**Comment by Etaash-mathamsetty** [comment]
> @Etaash-mathamsetty Still can't get it to detect EAC. I copied the EAC folder and placed it inside the `/Games/Heroic` folder, renamed it to something easier and the game still complains not finding it.
> 
> ![image](https://user-images.githubusercontent.com/39744079/175188296-3b9d61f7-1dd9-4d86-8558-7acc068e1f7b.png)
> 
> I'm using Proton-Experimental.

remove the quotes

**Comment by AlonsoCagi** [comment]
So I'll hijack this thread to ask for a similar solution but on Windows. From Heroic Launcher the game doesn't run, while executingFallGuys_client_game.exe by itself will open up untill the start screen and not any further with an EOS error. Any ideas?

If this isn't the correct instance to discuss my issues please let me know so I can open a different issue thread.

**Comment by 20PercentRendered** [comment]
I did all of the steps above, and once I got EAC loading from the runtime, the game crashed on the initial loading screen. Looking at the logs I get:
Error Log:
```
[cli] INFO: Logging in...
[Core] INFO: Trying to re-use existing login session...
[cli] INFO: Checking for updates...
[Core] INFO: Getting authentication token...
[cli] INFO: Launching 0a2d9f6403244d12969e11da6713137b...
fsync: up and running.
wine: Using setpriority to control niceness in the [-11,11] range
[S_API FAIL] SteamAPI_Init() failed; no appID found.
Either launch the game from Steam, or put the file steam_appid.txt containing the correct appID in your game folder.
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_S_construct null not valid
```
The last 2 lines are the culprit. But what is the issue?

**Comment by emmandyar** [comment]
Sorry to dissapear, had to run. Yeah, the quotes seems to be the issue, once they're gone, the game loaded! Currentply playing!

**Comment by Vanheelsing** [comment]
> > @Etaash-mathamsetty Still can't get it to detect EAC. I copied the EAC folder and placed it inside the `/Games/Heroic` folder, renamed it to something easier and the game still complains not finding it.
> > ![image](https://user-images.githubusercontent.com/39744079/175188296-3b9d61f7-1dd9-4d86-8558-7acc068e1f7b.png)
> > I'm using Proton-Experimental.
> 
> remove the quotes, I am glad I only spent 15 minutes figuring out the quotes were the problem, otherwise I would have never figured it out

So it's not possible to use EAC runtime that was installed via steam? I have also copied to heroic folder but the game still says the EAC is not initialized

**Comment by sheabert33** [comment]
> > > @Etaash-mathamsetty Still can't get it to detect EAC. I copied the EAC folder and placed it inside the `/Games/Heroic` folder, renamed it to something easier and the game still complains not finding it.
> > > ![image](https://user-images.githubusercontent.com/39744079/175188296-3b9d61f7-1dd9-4d86-8558-7acc068e1f7b.png)
> > > I'm using Proton-Experimental.
> > 
> > 
> > remove the quotes, I am glad I only spent 15 minutes figuring out the quotes were the problem, otherwise I would have never figured it out
> 
> So it's not possible to use EAC runtime that was installed via steam? I have also copied to heroic folder but the game still says the EAC is not initialized

I'm still getting this error as well.  I've followed all of the instructions in this thread so far but it doesn't seem to be working for me (yes, I removed the quotes in my launcher as well).

**Comment by Sultech** [comment]
After trying all methods, from reddit and here, I got it working by using the heroic deb package and using HeroicBashLauncher. I added the game to steam using HeroicBashLauncher, then I just used this launch option
 PROTON_EAC_RUNTIME="/home/(replace this with ur user)/.steam/steam/steamapps/common/Proton EasyAntiCheat Runtime/" %command%

Every time I tried to use the flatpak the game would crash with the launch options.

**Comment by Vanheelsing** [comment]
Good news
https://www.reddit.com/r/SteamDeck/comments/vht8d1/fall_guys_f2p_epic_version_tutorial_for_steam_deck/
Worked perfectly for me

**Comment by sheabert33** [comment]
> Good news https://www.reddit.com/r/SteamDeck/comments/vht8d1/fall_guys_f2p_epic_version_tutorial_for_steam_deck/ Worked perfectly for me

I ended up uninstalling from Heroic and reinstalling via Lutris.  Works perfectly now.  I'm not really sure why, but I was unable to get it working through Steam via Heroic.  It was probably something on my end, I'm a Linux noob for sure.  Lutris' instructions are much easier and shorter IMO.

**Comment by Etaash-mathamsetty** [comment]
> After trying all methods, from reddit and here, I got it working by using the heroic deb package and using HeroicBashLauncher. I added the game to steam using HeroicBashLauncher, then I just used this launch option PROTON_EAC_RUNTIME="/home/(replace this with ur user)/.steam/steam/steamapps/common/Proton EasyAntiCheat Runtime/" %command%
> 
> Every time I tried to use the flatpak the game would crash with the launch options.

no quotes

**Comment by Ahmed-E-86** [comment]
> > > > @Etaash-mathamsetty Still can't get it to detect EAC. I copied the EAC folder and placed it inside the `/Games/Heroic` folder, renamed it to something easier and the game still complains not finding it.
> > > > ![image](https://user-images.githubusercontent.com/39744079/175188296-3b9d61f7-1dd9-4d86-8558-7acc068e1f7b.png)
> > > > I'm using Proton-Experimental.
> > > 
> > > 
> > > remove the quotes, I am glad I only spent 15 minutes figuring out the quotes were the problem, otherwise I would have never figured it out
> > 
> > 
> > So it's not possible to use EAC runtime that was installed via steam? I have also copied to heroic folder but the game still says the EAC is not initialized
> 
> I'm still getting this error as well. I've followed all of the instructions in this thread so far but it doesn't seem to be working for me (yes, I removed the quotes in my launcher as well).

Here is what worked for me:

1- I installed `Proton EasyAntiCheat Runtime` using Steam desktop application.
2- I made a symlink to `~/.steam/debian-installation/steamapps/common/Proton EasyAntiCheat Runtime` _in the same folder_, but I replaced all **spaces** with **underscores** _to keep it updated_.
3- I used this env variable to launch the game `PROTON_EAC_RUNTIME=~/.steam/root/steamapps/common/Proton_EasyAntiCheat_Runtime`

I played multiple matches, and I did not encounter any issues, but I do not like this game.

**Comment by emmandyar** [comment]
I copied Steam's Protin EAC runtime and renamed it to something short so I
could write it down easily in Heroic's config. That got rid of  the spaces
in the folder's name and that's why it worked once I removed the quotation
marks

**Comment by LNDF** [comment]
> I did all of the steps above, and once I got EAC loading from the runtime, the game crashed on the initial loading screen. Looking at the logs I get: Error Log:
> 
> ```
> [cli] INFO: Logging in...
> [Core] INFO: Trying to re-use existing login session...
> [cli] INFO: Checking for updates...
> [Core] INFO: Getting authentication token...
> [cli] INFO: Launching 0a2d9f6403244d12969e11da6713137b...
> fsync: up and running.
> wine: Using setpriority to control niceness in the [-11,11] range
> [S_API FAIL] SteamAPI_Init() failed; no appID found.
> Either launch the game from Steam, or put the file steam_appid.txt containing the correct appID in your game folder.
> terminate called after throwing an instance of 'std::logic_error'
>   what():  basic_string::_S_construct null not valid
> ```
> 
> The last 2 lines are the culprit. But what is the issue?

Same issuse here. Steam is not un system fray...

**Comment by sheabert33** [comment]
> > > Good news https://www.reddit.com/r/SteamDeck/comments/vht8d1/fall_guys_f2p_epic_version_tutorial_for_steam_deck/ Worked perfectly for me
> > 
> > 
> > I ended up uninstalling from Heroic and reinstalling via Lutris. Works perfectly now. I'm not really sure why, but I was unable to get it working through Steam via Heroic. It was probably something on my end, I'm a Linux noob for sure. Lutris' instructions are much easier and shorter IMO.
> 
> I think I may need to go that route. I'm stuck trying to get it to work with Heroic, even after following a video guide. I'm a Linux noob as well and I have Lutris already. Was there a step-by-step guide you followed to get this to work with Lutris with EasyAntiCheat?

I found a pretty straight forward one on reddit.

https://www.reddit.com/r/linux_gaming/comments/vi91n6/guide_on_how_to_get_fall_guysepic_working_under/

**Comment by Ahmed-E-86** [comment]
The games stopped working for 2 days for no apparent reason. I found a similar game on Steam, its name is **Stumble Guys**, it is fine after trying it, and it does not use EAC.

**Comment by Etaash-mathamsetty** [comment]
> The game have stopped working for 2 days for no apparent reason. I found a similar game on Steam, its name is **Stumble Guys**, it is fine after trying it, and it does not use EAC.

try deleting your wine prefix and letting wine re-create it

**Comment by Ahmed-E-86** [comment]
> > The game have stopped working for 2 days for no apparent reason. I found a similar game on Steam, its name is **Stumble Guys**, it is fine after trying it, and it does not use EAC.
> 
> try deleting your wine prefix and letting wine re-create it the game still works great for me

I got it to work. Apparently the game only works with Wine-GE-Proton7-16, otherwise it will just freeze.

**Comment by BrianUribe6** [comment]
For anyone still having issues I made a script to automatically set up Fall Guys. 
It assumes you have Steam with Proton EasyAntiCheat Runtime installed. 

Tested on Manjaro:
[Link to Gist](https://gist.github.com/BrianUribe6/07ab372cdfdc8f6410e9d717ffcd1583)

**Comment by Etaash-mathamsetty** [comment]
this should become much easier once this gets merged
https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/1560

**Comment by Cris-lml007** [comment]
> > After trying all methods, from reddit and here, I got it working by using the heroic deb package and using HeroicBashLauncher. I added the game to steam using HeroicBashLauncher, then I just used this launch option PROTON_EAC_RUNTIME="/home/(replace this with ur user)/.steam/steam/steamapps/common/Proton EasyAntiCheat Runtime/" %command%
> > 
> > Every time I tried to use the flatpak the game would crash with the launch options.
> 
> no quotes pls, use the lutris easy anti cheat runtime, which has no spaces in its path (quotes make it not work for some reason)

what version of proton are you using?, i tried to use the eac from lutris and steam but when it opens it only loads for a while and closes, i'm using proton-ge-7.21

**Comment by Etaash-mathamsetty** [comment]
> > > After trying all methods, from reddit and here, I got it working by using the heroic deb package and using HeroicBashLauncher. I added the game to steam using HeroicBashLauncher, then I just used this launch option PROTON_EAC_RUNTIME="/home/(replace this with ur user)/.steam/steam/steamapps/common/Proton EasyAntiCheat Runtime/" %command%
> > > 
> > > Every time I tried to use the flatpak the game would crash with the launch options.
> > 
> > no quotes pls, use the lutris easy anti cheat runtime, which has no spaces in its path (quotes make it not work for some reason)
> 
> what version of proton are you using?, i tried to use the eac from lutris and steam but when it opens it only loads for a while and closes, i'm using proton-ge-7.21

What is loading for a while?
The eac popup or the actual game?

**Comment by Cris-lml007** [comment]
> > > > After trying all methods, from reddit and here, I got it working by using the heroic deb package and using HeroicBashLauncher. I added the game to steam using HeroicBashLauncher, then I just used this launch option PROTON_EAC_RUNTIME="/home/(replace this with ur user)/.steam/steam/steamapps/common/Proton EasyAntiCheat Runtime/" %command%
> > > > 
> > > > Every time I tried to use the flatpak the game would crash with the launch options.
> > > 
> > > no quotes pls, use the lutris easy anti cheat runtime, which has no spaces in its path (quotes make it not work for some reason)
> > 
> > what version of proton are you using?, i tried to use the eac from lutris and steam but when it opens it only loads for a while and closes, i'm using proton-ge-7.21
> 
> What is loading for a while?
> The eac popup or the actual game?

the actual game, the game pointer and the loading wheel appear in the lower right, but it only lasts about 3 seconds and the game closes by itself.

**Comment by Nocccer** [comment]
Guide that works without gamemode and without steam:

- Follow the "Editing the files" part of https://steamcommunity.com/sharedfiles/filedetails/?id=2787044978 (move the .so file and edit the .ini. It will not work with just the .so file)
- Download the EAC runtime from https://github.com/lutris/buildbot/releases/download/2021.11.20/eac_runtime.tar.xz, move the file into ~/Games/Heroic/ and extract it (so you have a eac_runtime folder at ~/Games/Heroic/ containing a v2 folder)
- Go into the game settings in Heroic -> "Other" -> "Advanced Options (Environment Variables)". Add the following: PROTON_EAC_RUNTIME=/home/YourUsername/Games/Heroic/eac_runtime LD_PRELOAD="". Replace YourUsername with your actual username

If things aren't working:
- Make sure the most recent version of Wine-GE is installed and selected as your Wine Version in the game settings in Heroic. It might also work with Proton, but Wine is more reliable and recommended
- Make sure your system is fully up to date and you followed the Wine Dependency guide at https://www.gloriouseggroll.tv/how-to-get-out-of-wine-dependency-hell/ (Bearbeitet)
Springen

**Comment by 20PercentRendered** [comment]
> 

If the game crashes on the initial loading screen it means you have set the EAC runtime successfully. However, you will need to run it from steam using HeroicBashLauncher, since the EAC runtime seems to depend on the Steam Linux Runtime.

**Comment by Nocccer** [comment]
> e EAC runtime seems to depend on the Steam Linux Runtime.

Wrong. The problem was that `LD_PRELOAD` was undefined. Defining it with a empty string fixes it. I got Fall Guys working with the guide described above. No GameMode, No Steam. Wine-GE-20

**Comment by nothor** [comment]
Tested with Fall Guys on Ubuntu 22.04 with last Heroic-2.4.0-beta.2 as Appimage with the feature EAC implemented and it worked including EOS Overlay
1) Unistall previous install including WinePrefix
2) Install "Wine-GE-Proton7-20" Wine Version
3) Install again Fall Guys and link it to "Wine-GE-Proton7-20"
4) Go to Fall Guys --> Tools --> Enable EOS Overlay
5) Check EasyAntiCheat inside Fall Guys --> Config --> Wine Extensions (by default is active)
6) Symlink "easyanticheat_x64.so" inside /Plugins/x86_64
7) Change TargetApplicationPath inside "FallGuys_client.ini" to _FallGuys_client_game.exe_

**Important:** In my case with an Intel/NVIDIA GPU I must uncheck Use Dedicated Graphic Card inside Fall Guys --> Config --> Other

**Comment by CommandMC** [comment]
Implemented in #1560
